src/lib/yang/state.lua

```diff
@@ -0,0 +1,92 @@
+-- Use of this source code is governed by the Apache 2.0 license; see COPYING.
+module(..., package.seeall)
+
+local lib = require("core.lib")
+local shm = require("core.shm")
+local xpath = require("lib.yang.xpath")
+local yang = require("lib.yang.yang")
+local yang_data = require("lib.yang.data")
+local counter = require("core.counter")
+
+local counter_directory = "/apps"
+
+local function find_counters(pid)
+    local path = shm.root.."/"..pid..counter_directory
+    local counters = {}
+    for _, c in pairs(lib.files_in_directory(path)) do
+        local counterdir = "/"..pid..counter_directory.."/"..c
+        counters[c] = shm.open_frame(counterdir)
+    end
+    return counters
+end
+
+function collect_state_leaves(schema)
+    -- Iterate over schema looking fo state leaves at a specific path into the
+    -- schema. This should return a dictionary of leaf to lua path.
+    local function collection(scm, path, config)
+        local function newpath(oldpath)
+            return lib.deepcopy(oldpath)
+        end
+        if path == nil then path = {} end
+
+        -- Add the current schema node to the path
+        table.insert(path, scm.id)
+
+        if scm.config ~= nil then
+            config = scm.config
+        end
+
+        if scm.kind == "container" then
```

Comment:
Probably we should use something like the grammar_from_schema thing; wdyt?  Maybe not tho.